File: src/main/java/com/example/legacyshop/service/OrderService.java
package com.example.legacyshop.service;

import com.example.legacyshop.dto.OrderCreateRequest;
import com.example.legacyshop.dto.OrderResponse;
import com.example.legacyshop.entity.*;
import com.example.legacyshop.exception.BusinessValidationException;
import com.example.legacyshop.exception.ResourceNotFoundException;
import com.example.legacyshop.repository.*;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.Optional;

/**
 * Service for Order operations demonstrating the complete order placement pipeline.
 * 
 * Key patterns demonstrated:
 * 1. Atomic transaction boundary
 * 2. Validation → Stock Check → Price Calculation → Persist → Stock Decrement
 * 3. Rollback on any failure
 * 4. Idempotency support
 * 5. Business rule enforcement
 */
@Service
@Transactional(readOnly = true)
public class OrderService {

    private final OrderRepository orderRepository;
    private final ProductRepository productRepository;
    private final CustomerRepository customerRepository;
    private final AuditLogRepository auditLogRepository;
    private final IdempotencyRecordRepository idempotencyRecordRepository;
    private final CustomerService customerService;
    private final DiscountService discountService;
    private final PaymentService paymentService;

    public OrderService(OrderRepository orderRepository,
                       ProductRepository productRepository,
                       CustomerRepository customerRepository,
                       AuditLogRepository auditLogRepository,
                       IdempotencyRecordRepository idempotencyRecordRepository,
                       CustomerService customerService,
                       DiscountService discountService,
                       PaymentService paymentService) {
        this.orderRepository = orderRepository;
        this.productRepository = productRepository;
        this.customerRepository = customerRepository;
        this.auditLogRepository = auditLogRepository;
        this.idempotencyRecordRepository = idempotencyRecordRepository;
        this.customerService = customerService;
        this.discountService = discountService;
        this.paymentService = paymentService;
    }

    /**
     * Create order with complete pipeline: validate → stock check → price → persist → decrement stock.
     * Demonstrates atomic rollback on any failure.
     */
    @Transactional
    public OrderResponse createOrder(OrderCreateRequest request, String idempotencyKey) {
        // Step 1: Check for idempotency (prevent duplicates)
        if (idempotencyKey != null) {
            Optional<OrderEntity> existingOrder = orderRepository.findByIdempotencyKey(idempotencyKey);
            if (existingOrder.isPresent()) {
                return OrderResponse.from(existingOrder.get());
            }
        }

        // Step 2: Validate customer (find or create)
        Customer customer = customerService.findOrCreateCustomer(
            request.getCustomerEmail(),
            extractFirstName(request.getCustomerEmail()),
            extractLastName(request.getCustomerEmail())
        );

        // Step 3: Create order entity
        OrderEntity order = new OrderEntity(customer, idempotencyKey);

        // Step 4: Process each line item with validation and stock checks
        for (OrderCreateRequest.OrderItemRequest itemRequest : request.getItems()) {
            processOrderItem(order, itemRequest);
        }

        // Step 5: Calculate totals and apply discounts
        order.calculateTotals();
        BigDecimal discountAmount = discountService.calculateDiscount(order.getSubtotal());
        order.applyDiscount(discountAmount);

        // Step 6: Validate minimum order total
        if (order.getTotal().compareTo(new BigDecimal("0.01")) < 0) {
            throw new BusinessValidationException("Order total must be at least $0.01");
        }

        // Step 7: Persist order
        OrderEntity savedOrder = orderRepository.save(order);

        // Step 8: Record idempotency (if provided)
        if (idempotencyKey != null) {
            IdempotencyRecord record = new IdempotencyRecord(idempotencyKey, "ORDER_CREATE");
            record.setResult(savedOrder.getId(), savedOrder.getStatus().name());
            idempotencyRecordRepository.save(record);
        }

        // Step 9: Create audit log
        AuditLog auditLog = AuditLog.orderCreatedLog(savedOrder.getId(), customer.getEmail());
        auditLogRepository.save(auditLog);

        return OrderResponse.from(savedOrder);
    }

    /**
     * Process individual order item with validation and atomic stock decrement.
     */
    private void processOrderItem(OrderEntity order, OrderCreateRequest.OrderItemRequest itemRequest) {
        // Find product by SKU
        Product product = productRepository.findBySku(itemRequest.getProductSku())
            .orElseThrow(() -> new ResourceNotFoundException("Product not found with SKU: " + itemRequest.getProductSku()));

        // Validate product is active
        if (!product.getActive()) {
            throw new BusinessValidationException("Product " + product.getSku() + " is not available");
        }

        // Check stock availability BEFORE decrementing
        if (product.getStockQuantity() < itemRequest.getQuantity()) {
            throw new BusinessValidationException(
                String.format("Insufficient stock for product %s. Available: %d, Requested: %d",
                    product.getSku(), product.getStockQuantity(), itemRequest.getQuantity())
            );
        }

        // Capture current price (price at order time, not derived)
        BigDecimal unitPrice = product.getPrice();

        // Create order item
        OrderItem orderItem = new OrderItem(order, product, itemRequest.getQuantity(), unitPrice);
        order.addItem(orderItem);

        // ATOMIC OPERATION: Decrement stock
        // This is the critical part - if this fails, entire transaction rolls back
        product.decrementStock(itemRequest.getQuantity());
        productRepository.save(product);
    }

    /**
     * Get order by ID.
     */
    public OrderResponse getOrder(Long id) {
        OrderEntity order = orderRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Order not found with id: " + id));
        return OrderResponse.from(order);
    }

    /**
     * Get orders for customer with pagination.
     */
    public Page<OrderResponse> getCustomerOrders(String email, Pageable pageable) {
        Customer customer = customerService.findByEmail(email);
        return orderRepository.findByCustomerId(customer.getId(), pageable)
            .map(OrderResponse::from);
    }

    /**
     * Cancel order with compensating actions (restore stock, void payment).
     */
    @Transactional
    public OrderResponse cancelOrder(Long orderId) {
        OrderEntity order = orderRepository.findById(orderId)
            .orElseThrow(() -> new ResourceNotFoundException("Order not found with id: " + orderId));

        if (!order.canBeCancelled()) {
            throw new BusinessValidationException("Order cannot be cancelled in status: " + order.getStatus());
        }

        // Compensating action 1: Restore stock for all items
        for (OrderItem item : order.getItems()) {
            Product product = item.getProduct();
            product.incrementStock(item.getQuantity());
            productRepository.save(product);
        }

        // Compensating action 2: Void payment if exists
        if (order.getPayment() != null && order.getPayment().getStatus() == Payment.PaymentStatus.AUTHORIZED) {
            paymentService.voidPayment(order.getPayment().getId());
        }

        // Update order status
        order.cancel();
        OrderEntity savedOrder = orderRepository.save(order);

        // Audit logging
        AuditLog auditLog = AuditLog.orderCancelledLog(savedOrder.getId(), "Customer requested cancellation");
        auditLogRepository.save(auditLog);

        return OrderResponse.from(savedOrder);
    }

    // Helper methods for customer name extraction (simplified)
    private String extractFirstName(String email) {
        String localPart = email.substring(0, email.indexOf('@'));
        return localPart.replaceAll("[^a-zA-Z]", "");
    }

    private String extractLastName(String email) {
        return "Customer"; // Simplified - in real app might extract from email or require in request
    }
}

File: legacyshop-api/app/services/order_service.py
"""
Order service implementing the complete order placement pipeline.
Mirrors the Java OrderService implementation.
"""

from sqlalchemy.orm import Session
from decimal import Decimal
from typing import Optional

from app.models.models import (
    Order, OrderItem, Customer, Product, IdempotencyRecord, AuditLog,
    OrderStatus, Payment, PaymentStatus
)
from app.schemas.schemas import OrderCreateRequest, OrderResponse, OrderItemResponse
from app.services.discount_service import discount_service
from app.services.payment_service import payment_service
from app.services.exceptions import BusinessValidationException, ResourceNotFoundException


class OrderService:
    """Service for Order operations demonstrating the complete order placement pipeline"""
    
    def create_order(self, request: OrderCreateRequest, idempotency_key: Optional[str], db: Session) -> OrderResponse:
        """
        Create order with complete pipeline: validate → stock check → price → persist → decrement stock.
        Demonstrates atomic rollback on any failure.
        """
        if idempotency_key:
            existing_order = db.query(Order).filter(Order.idempotency_key == idempotency_key).first()
            if existing_order:
                return self._order_to_response(existing_order)
        
        customer = self._find_or_create_customer(request.customerEmail, db)
        
        order = Order(
            customer_id=customer.id,
            idempotency_key=idempotency_key,
            status=OrderStatus.PENDING
        )
        db.add(order)
        db.flush()  # Get order ID without committing
        
        for item_request in request.items:
            self._process_order_item(order, item_request, db)
        
        db.flush()
        db.refresh(order)
        
        order.calculate_totals()
        discount_amount = discount_service.calculate_discount(Decimal(str(order.subtotal)))
        order.apply_discount(float(discount_amount))
        
        if order.total < 0.01:
            raise BusinessValidationException("Order total must be at least $0.01")
        
        db.commit()
        db.refresh(order)
        
        if idempotency_key:
            idempotency_record = IdempotencyRecord(
                idempotency_key=idempotency_key,
                operation_type="ORDER_CREATE",
                result_entity_id=order.id,
                result_status=order.status.value
            )
            db.add(idempotency_record)
        
        audit_log = AuditLog(
            entity_type="ORDER",
            entity_id=order.id,
            action="CREATED",
            details=f"Order created for customer {customer.email}"
        )
        db.add(audit_log)
        db.commit()
        
        return self._order_to_response(order)
    
    def _process_order_item(self, order: Order, item_request, db: Session):
        """Process individual order item with validation and atomic stock decrement"""
        product = db.query(Product).filter(Product.sku == item_request.productSku).first()
        if not product:
            raise ResourceNotFoundException(f"Product not found with SKU: {item_request.productSku}")
        
        if not product.active:
            raise BusinessValidationException(f"Product {product.sku} is not available")
        
        if product.stock_quantity < item_request.quantity:
            raise BusinessValidationException(
                f"Insufficient stock for product {product.sku}. "
                f"Available: {product.stock_quantity}, Requested: {item_request.quantity}"
            )
        
        unit_price = product.price
        
        order_item = OrderItem(
            order_id=order.id,
            product_id=product.id,
            product_sku=product.sku,
            product_name=product.name,
            quantity=item_request.quantity,
            unit_price=unit_price,
            subtotal=round(unit_price * item_request.quantity, 2)
        )
        db.add(order_item)
        
        product.decrement_stock(item_request.quantity)
    
    def _find_or_create_customer(self, email: str, db: Session) -> Customer:
        """Find existing customer or create new one"""
        customer = db.query(Customer).filter(Customer.email == email).first()
        if not customer:
            local_part = email.split('@')[0]
            first_name = ''.join(c for c in local_part if c.isalpha())
            
            customer = Customer(
                email=email,
                first_name=first_name or "Customer",
                last_name="Customer"
            )
            db.add(customer)
            db.flush()
        return customer
    
    def get_order(self, order_id: int, db: Session) -> OrderResponse:
        """Get order by ID"""
        order = db.query(Order).filter(Order.id == order_id).first()
        if not order:
            raise ResourceNotFoundException(f"Order not found with id: {order_id}")
        return self._order_to_response(order)
    
    def get_customer_orders(self, email: str, page: int, size: int, db: Session):
        """Get orders for customer with pagination"""
        customer = db.query(Customer).filter(Customer.email == email).first()
        if not customer:
            raise ResourceNotFoundException(f"Customer not found with email: {email}")
        
        query = db.query(Order).filter(Order.customer_id == customer.id).order_by(Order.created_at.desc())
        total = query.count()
        orders = query.offset(page * size).limit(size).all()
        
        return {
            "content": [self._order_to_response(order) for order in orders],
            "totalElements": total,
            "totalPages": (total + size - 1) // size,
            "size": size,
            "number": page
        }
    
    def cancel_order(self, order_id: int, db: Session) -> OrderResponse:
        """Cancel order with compensating actions (restore stock, void payment)"""
        order = db.query(Order).filter(Order.id == order_id).first()
        if not order:
            raise ResourceNotFoundException(f"Order not found with id: {order_id}")
        
        if not order.can_be_cancelled():
            raise BusinessValidationException(f"Order cannot be cancelled in status: {order.status.value}")
        
        for item in order.items:
            product = db.query(Product).filter(Product.id == item.product_id).first()
            if product:
                product.increment_stock(item.quantity)
        
        if order.payment and order.payment.status == PaymentStatus.AUTHORIZED:
            payment_service.void_payment(order.payment.id, db)
        
        order.cancel()
        
        audit_log = AuditLog(
            entity_type="ORDER",
            entity_id=order.id,
            action="CANCELLED",
            details="Customer requested cancellation"
        )
        db.add(audit_log)
        
        db.commit()
        db.refresh(order)
        
        return self._order_to_response(order)
    
    def _order_to_response(self, order: Order) -> OrderResponse:
        """Convert Order entity to OrderResponse"""
        items = [
            OrderItemResponse(
                id=item.id,
                productSku=item.product_sku,
                productName=item.product_name,
                quantity=item.quantity,
                unitPrice=item.unit_price,
                subtotal=item.subtotal
            )
            for item in order.items
        ]
        
        return OrderResponse(
            id=order.id,
            customerEmail=order.customer.email,
            status=order.status.value,
            subtotal=order.subtotal,
            discountAmount=order.discount_amount,
            total=order.total,
            idempotencyKey=order.idempotency_key,
            items=items,
            createdAt=order.created_at,
            updatedAt=order.updated_at
        )


order_service = OrderService()
